### PR TITLE
Drop unidecode and replace with cautious_slugify from Wagtail

### PR DESCRIPTION
**Pull request by craigloftus**
Unidecode's license is GPLv2, which is not compatible with Oscar's license. See #2745 for discussion.

This PR introduces the `cautious_slugify` function from Wagtail as an alternative to produce non-empty slugs from unicode strings. See the tests for examples of the change in output.

I changed the API slightly to drop support for directly passing the `OSCAR_SLUG_FUNCTION` a function and also changed the call signature of the slug function to match the Django slugify function; which will make it easier for downstream projects to decide how to handle unicode.

### CODE REVIEW

File: src/oscar/core/utils.py [hunk 1]
```diff
@@ -8,43 +10,89 @@
...
-    Slugify a string (even if it contains non-ASCII chars)
+    Slugify a string
+
+    The OSCAR_SLUG_FUNCTION can be set with a dotted path to the slug
+    function to use, defaults to 'oscar.core.utils.default_slugifier'.
+
+    OSCAR_SLUG_MAP can be set of a dictionary of target:replacement pairs
+
+    OSCAR_SLUG_BLACKLIST can be set to a iterable of words to remove after
+    the slug is generated; though it will not reduce a slug to zero length.
```
Comment (solarissmoke) [inline src/oscar/core/utils.py]: Think we need a test for this behaviour ("it will not reduce a slug to zero length").

File: docs/source/releases/v2.0.rst [hunk 2]
```diff
@@ -40,6 +40,14 @@ Minor changes
 - ``OrderPlacementMixin.get_message_context()`` is now passed a ``code`` argument
   specifying the communication event type code for a message being sent.
 
+- We've dropped the dependency on Unidecode due to license incompatibilities,
+  ``oscar.core.utils.default_slugifier`` now uses ``oscar.core.utils.cautious_slugify``
+  to handle unicode characters in slugs when ``OSCAR_SLUG_ALLOW_UNICODE`` is ``False``.
+
+- ``OSCAR_SLUG_FUNCTION`` previously accepted a function as its value, it now
```
Comment (solarissmoke) [inline docs/source/releases/v2.0.rst]: `It now only accepts...` should be a new sentence.

File: src/oscar/core/utils.py [hunk 3]
```diff
@@ -8,43 +10,89 @@
...
+
+def default_slugifier(value, allow_unicode=False):
     """
-    Oscar's default slugifier function. Uses Django's slugify function.
+    Oscar's default slugifier function. When unicode is allowed
+    it uses Django's slugify function, otherwise it uses cautious_slugify.
     """
-    return django_slugify(value, allow_unicode=settings.OSCAR_SLUG_ALLOW_UNICODE)
+    if allow_unicode:
+        slug = django_slugify(value, allow_unicode=True)
```
Comment (solarissmoke) [inline src/oscar/core/utils.py]: Minor, but you could just return here instead of assigning to `slug`.

File: src/oscar/core/utils.py (hunk 1 above)
Comment (craigloftus) [reply src/oscar/core/utils.py]: That is what `test_uses_blacklist_all_words` is testing for

Comment (solarissmoke) [reply src/oscar/core/utils.py]: As yes, OK.
